services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/handlers/CopySourceInterceptor.java

```diff
@@ -0,0 +1,127 @@
+            requireNotSet(request.sourceBucket(), "sourceBucket");
+            requireNotSet(request.sourceKey(), "sourceKey");
+            requireNotSet(request.sourceVersionId(), "sourceVersionId");
+            return request;
+        }
+        String copySource = constructCopySource(
+            requireSet(request.sourceBucket(), "sourceBucket"),
+            requireSet(request.sourceKey(), "sourceKey"),
+            request.sourceVersionId()
+        );
+        return request.toBuilder()
+                      .sourceBucket(null)
+                      .sourceKey(null)
+                      .sourceVersionId(null)
+                      .copySource(copySource)
+                      .build();
+    }
+
+    private static SdkRequest modifyUploadPartCopyRequest(UploadPartCopyRequest request) {
+        if (request.copySource() != null) {
+            requireNotSet(request.sourceBucket(), "sourceBucket");
+            requireNotSet(request.sourceKey(), "sourceKey");
+            requireNotSet(request.sourceVersionId(), "sourceVersionId");
+            return request;
+        }
+        String copySource = constructCopySource(
+            requireSet(request.sourceBucket(), "sourceBucket"),
+            requireSet(request.sourceKey(), "sourceKey"),
+            request.sourceVersionId()
+        );
+        return request.toBuilder()
+                      .sourceBucket(null)
+                      .sourceKey(null)
+                      .sourceVersionId(null)
+                      .copySource(copySource)
+                      .build();
+    }
+
+    private static String constructCopySource(String sourceBucket, String sourceKey, String sourceVersionId) {
+        StringBuilder copySource = new StringBuilder();
+        copySource.append("/");
+        copySource.append(urlEncodeIgnoreSlashes(sourceBucket));
+        if (isArnFor(S3ResourceType.ACCESS_POINT, sourceBucket) || isArnFor(S3ResourceType.OUTPOST, sourceBucket)) {
+            copySource.append("/object");
+        }
+        copySource.append("/");
+        copySource.append(urlEncodeIgnoreSlashes(sourceKey));
+        if (sourceVersionId != null) {
+            copySource.append("?versionId=");
+            copySource.append(urlEncodeIgnoreSlashes(sourceVersionId));
+        }
+        return copySource.toString();
+    }
+
+    private static void requireNotSet(Object value, String paramName) {
+        if (value != null) {
+            throw new IllegalArgumentException(String.format("Parameter 'copySource' must not be used in conjunction with '%s'",
+                                                             paramName));
+        }
+    }
+
+    private static <T> T requireSet(T value, String paramName) {
+        if (value == null) {
+            throw new IllegalArgumentException(String.format("Parameter '%s' must not be null",
+                                                             paramName));
+        }
+        return value;
```

Comment:
Nit: You could have used `Validate.isTrue` here. I don't think it's much easier to read, so no real need to change this unless you want to.

Reply:
Will change. I like not having to specify `String.format(..)`.